### PR TITLE
Add mallet CRF Tagger

### PR DESCRIPTION
**Pull request by longdt219**
Add mallet CRF tagger #104

The accuracy for CRF Tagger evaluated on part of English Penn Treebank 
Training data : 300 sentences 
Testing data : 214 sentences 
--> Accuracy : 81 % 

I also updated the instruction for using Mallet on Wiki.

### CODE REVIEW

File: nltk/tag/mallet.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+from nltk.tag.api import TaggerI
+
+_mallet_url = 'http://mallet.cs.umass.edu/grmm/'
+
+class MalletTagger(TaggerI):
+    """
+    An interface to Mallet CRF taggers.
+    
+    >>> from nltk.tag.mallet import MalletTagger
+    >>> mt = MalletTagger('/Users/HeroAthen/Dropbox/Work/nltk/temp/mallet-2.0.7/dist/mallet.jar',
```
Comment (kmike) [inline nltk/tag/mallet.py:29]: The test shouldn't depend on hardcoded Mallet paths - it will fail on CI server and on users' machines.

File: nltk/tag/mallet.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            
+    def _get_features(self,data):
+        """
+        Extract basic features about this word including 
+             - Current Word 
+             - Is Capitalized ?
+             - Has Punctuation ?
+             - Has Number ?
+             - Suffixes up to length 3 
+        :return : a string which contains the features
```
Comment (kmike) [inline nltk/tag/mallet.py:80]: Is it an example Mallet wrapper for POS tagging? If so, where did the existing features came from? It could make sense to use features from some published paper, e.g. from http://www.inf.ed.ac.uk/resources/nlp/local_doc/mxpost_doc.pdf, with a reference. 

Alternatively, if it is not just an educational example, and if it is intended to be a production-quality POS tagger, we can make model use features which gives good results on a large treebank. 81% accuracy on POS tagging is far from the state of the arts; it may be explained by a small training data size (300/214 sentences is not a large treebank), but if so features shouldn't be chosen based on what gives best results on this dataset.

File: nltk/tag/mallet.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+             - Current Word 
+             - Is Capitalized ?
+             - Has Punctuation ?
+             - Has Number ?
+             - Suffixes up to length 3 
+        :return : a string which contains the features   
+        
+        >>> from nltk.tag.mallet import MalletTagger
+        >>> mt = MalletTagger()
+        >>> mt._get_features(u'University1')
```
Comment (kmike) [inline nltk/tag/mallet.py:84]: this doctest will fail in Python 3.2

File: nltk/tag/mallet.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        :rtype : list (list (tuple(str,str))) 
+        '''
+        
+        # We need the list of sentences instead of the list generator for matching the input and output  
+        sentences = list(sents) 
+        
+        # First, build the test file 
+        input_file = tempfile.NamedTemporaryFile(
+                prefix='mallet_tagger.test',
+                dir=tempfile.gettempdir(),
```
Comment (kmike) [inline nltk/tag/mallet.py:131]: dir argument doesn't look necessary here

File: nltk/tag/mallet.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                delete=False)
+
+        for sent in sentences: 
+            for token in sent:
+                data = unicode(token)
```
Comment (kmike) [inline nltk/tag/mallet.py:136]: What is this line for? If the input is unicode then unicode(token) does nothing; if it is not unicode then the text is decoded using some system-depending encoding which is not good. And there is no "unicode" symbol in Python 3.x, so it doesn't work in Python 3.x

File: nltk/tag/mallet.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+                input_file.write(self._get_features(data) + '\n')
```
Comment (kmike) [inline nltk/tag/mallet.py:137]: Unicode handling is broken here: `self._get_features` returns unicode, but input_file is open in binary mode.

**Comment by kmike** [comment]
Hey @longdt219,

Thanks for the contribution!

My main question is why are we going with Mallet for linear-chain CRFs? Some of the problems with Mallet:
1. Mallet is not a Python package, it should be installed and configured. For one, it makes it harder to write unit tests.
2. It seems that Mallet is slower than other linear-chain CRF implementations. For example, here http://www.chokkan.org/software/crfsuite/benchmark.html author of CRFSuite claims Mallet is up to 100x slower than CRFSuite; also, as implemented, the tagger starts Mallet binary each time `tag_sents` is called.

Another question is what is this wrapper for. Now it looks like an example of how to write a POS tagger using Mallet. Are users supposed to define their own features? If so, the I think this interface should be public, and the example POS tagger class should be clearly separated from the Mallet wrapper itself.

**Comment by stevenbird** [comment]
Thanks for the detailed feedback @kmike. I see that you're behind the [Python interface to CRFSuite](https://pypi.python.org/pypi/python-crfsuite/). The only reason to support Mallet is if there is other functionality that people want (like topic modelling, [mentioned by @ewan-klein](https://github.com/nltk/nltk/issues/104#issuecomment-34634881)) or existing trained models that people want to use (for replicating experiments). However, it looks like [there hasn't been any interest shown in Mallet for NLTK in the past 18 months](https://www.google.com/webhp?#q=mallet+nltk), so I would be happy to use CRFSuite for CRF tagging instead.

Separately from this, we can always come back to using Mallet for topic modelling, after considering the alternatives.

Doesn't your objection 1 above apply equally to CRFSuite, @kmike?

File: nltk/tag/mallet.py (hunk 1 above)
Comment (longdt219) [inline nltk/tag/mallet.py:29]: If mallet paths are not found, it will look at the environment variable MALLET and MALLET_DEPS.

Comment (kmike) [inline nltk/tag/mallet.py:29]: Aha, I see, thanks for an explanation.

**Comment by kmike** [comment]
Hi @stevenbird,

Yes, https://github.com/tpeng/python-crfsuite is one alternative for linear-chain CRFs; others are https://github.com/adsva/python-wapiti and https://github.com/larsmans/seqlearn, and maybe there are some other packages. Both python-crfsuite and python-wapiti bundle CRFsuite and Wapiti libraries, so users get them just by running 'pip install python-crfsuite' or 'pip install libwapiti'. There is no need to install CRFsuite or Wapiti separately. SeqLearn doesn't bundle any external non-Python libraries, but it requires numpy and scikit-learn (also, it doesn't implement CRFs, but there are similar models in seqlearn, and a PR for CRF training).

All of these libraries are based on Python C/C++ extensions. It means that while they can be installed with pip, users must have a working build chain on their machines (a compiler, etc). I'm not sure if it is a problem; many other Python packages need a working build environment to be installed, and this environment is relatively easy to setup unless you're on Windows. 

As for the topic modelling, gensim is a popular choice in Python. It has lots of useful stuff built-in like LSI or word2vec implementation or doc2vec (Quoc Le and Tomas Mikolov. Distributed Representations of Sentences and Documents) implementation. And it also has a wrapper for Mallet LDA, so by using gensim we're getting a wrapper for Mallet topic modelling with the same interface as other gensim models.

**Comment by stevenbird** [comment]
Thanks @kmike. I'm happy to presume that anyone doing substantial machine learning tasks has a functioning build environment. Given the simplicity of installing CRFSuite, and the better speed than Mallet, I think we should just go with CRFSuite.

To interface with gensim, I think we just need to provide a Matrix Market view on NLTK corpora. Would that do it?

**Comment by longdt219** [comment]
Hi @stevenbird,
Should I write a wrapper for CRFSuite. I don't think it is necessary given that <a href='https://github.com/tpeng/python-crfsuite' > someone did it for  python  </a> and the <a href='http://nbviewer.ipython.org/github/tpeng/python-crfsuite/blob/master/examples/CoNLL%202002.ipynb'> usage </a> is fairy straight forward.

**Comment by stevenbird** [comment]
@longdt219, the idea is to have an implementation of `TaggerI` using CRFSuite, so that someone can switch between different taggers just by a change of import statement.

**Comment by stevenbird** [comment]
We've agreed to use CRFSuite instead. The main advantage of Mallet is configurability, and this is not exposed via the NLTK wrapper. We prefer the faster speed of CRFSuite.
